"""
gRPCå®Œæ•´åŠŸèƒ½æ¼”ç¤º
å±•ç¤ºå¦‚ä½•ä½¿ç”¨æ‰€æœ‰gRPCæœåŠ¡åŠŸèƒ½
"""

import grpc
import sys
import os
import time

# æ·»åŠ å½“å‰ç›®å½•åˆ°Pythonè·¯å¾„
sys.path.append(os.path.dirname(os.path.abspath(__file__)))
sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

# å¯¼å…¥ç”Ÿæˆçš„gRPCä»£ç 
import role_play_pb2
import role_play_pb2_grpc

def demo_all_features():
    """æ¼”ç¤ºæ‰€æœ‰gRPCåŠŸèƒ½"""
    # è¿æ¥åˆ°gRPCæœåŠ¡å™¨
    with grpc.insecure_channel('localhost:50051') as channel:
        stub = role_play_pb2_grpc.RolePlayServiceStub(channel)
        
        print("=== AIè§’è‰²æ‰®æ¼”gRPCåŠŸèƒ½æ¼”ç¤º ===\n")
        
        # 1. è·å–é¢„è®¾è§’è‰²åˆ—è¡¨
        print("1. è·å–é¢„è®¾è§’è‰²åˆ—è¡¨")
        try:
            response = stub.GetPresetRoles(role_play_pb2.GetPresetRolesRequest())
            print(f"   æˆåŠŸè·å–åˆ° {len(response.roles)} ä¸ªé¢„è®¾è§’è‰²:")
            for i, role in enumerate(response.roles[:5]):  # æ˜¾ç¤ºå‰5ä¸ª
                print(f"   {i+1}. {role.emoji} {role.name}")
            if len(response.roles) > 5:
                print(f"   ... è¿˜æœ‰ {len(response.roles) - 5} ä¸ªè§’è‰²")
            print()
        except Exception as e:
            print(f"   è·å–è§’è‰²åˆ—è¡¨å¤±è´¥: {e}\n")
            return
        
        # 2. è®¾ç½®è§’è‰²
        print("2. è®¾ç½®è§’è‰²ä¸º 'ğŸ§™â€â™‚ï¸ æ™ºæ…§å¯¼å¸ˆ'")
        try:
            response = stub.SetRole(role_play_pb2.SetRoleRequest(
                role_name="ğŸ§™â€â™‚ï¸ æ™ºæ…§å¯¼å¸ˆ"
            ))
            if response.success:
                print(f"   è§’è‰²è®¾ç½®æˆåŠŸ!\n")
            else:
                print(f"   è§’è‰²è®¾ç½®å¤±è´¥: {response.message}\n")
        except Exception as e:
            print(f"   è®¾ç½®è§’è‰²å¤±è´¥: {e}\n")
            return
        
        # 3. ä¸è§’è‰²å¯¹è¯
        print("3. ä¸è§’è‰²å¯¹è¯")
        try:
            response = stub.ChatWithRole(role_play_pb2.ChatRequest(
                message="ä½ å¥½ï¼Œèƒ½ç®€å•ä»‹ç»ä¸€ä¸‹ä»€ä¹ˆæ˜¯äººå·¥æ™ºèƒ½å—ï¼Ÿ"
            ))
            if response.success:
                print(f"   AIå›å¤: {response.response[:100]}...\n")  # åªæ˜¾ç¤ºå‰100ä¸ªå­—ç¬¦
            else:
                print(f"   å¯¹è¯å¤±è´¥: {response.error}\n")
        except Exception as e:
            print(f"   å¯¹è¯å¤±è´¥: {e}\n")
        
        # 4. ç»§ç»­å¯¹è¯
        print("4. ç»§ç»­å¯¹è¯")
        try:
            response = stub.ChatWithRole(role_play_pb2.ChatRequest(
                message="é‚£æœºå™¨å­¦ä¹ å’Œæ·±åº¦å­¦ä¹ æœ‰ä»€ä¹ˆåŒºåˆ«å‘¢ï¼Ÿ"
            ))
            if response.success:
                print(f"   AIå›å¤: {response.response[:100]}...\n")  # åªæ˜¾ç¤ºå‰100ä¸ªå­—ç¬¦
            else:
                print(f"   å¯¹è¯å¤±è´¥: {response.error}\n")
        except Exception as e:
            print(f"   å¯¹è¯å¤±è´¥: {e}\n")
        
        # 5. é‡ç½®å¯¹è¯
        print("5. é‡ç½®å¯¹è¯")
        try:
            response = stub.ResetConversation(role_play_pb2.ResetRequest())
            if response.success:
                print(f"   å¯¹è¯é‡ç½®æˆåŠŸ: {response.message}\n")
            else:
                print(f"   é‡ç½®å¤±è´¥: {response.message}\n")
        except Exception as e:
            print(f"   é‡ç½®å¤±è´¥: {e}\n")
        
        # 6. è®¾ç½®è‡ªå®šä¹‰è§’è‰²
        print("6. è®¾ç½®è‡ªå®šä¹‰è§’è‰²")
        try:
            response = stub.SetRole(role_play_pb2.SetRoleRequest(
                role_name="ğŸ­ è‡ªå®šä¹‰è§’è‰²",
                custom_name="å†å²å­¦å®¶",
                custom_prompt="ä½ æ˜¯ä¸€ä½èµ„æ·±çš„å†å²å­¦å®¶ï¼Œä¸“é—¨ç ”ç©¶ä¸­å›½å¤ä»£å²ã€‚ä½ çŸ¥è¯†æ¸Šåšï¼Œå–„äºç”¨ç”ŸåŠ¨çš„ä¾‹å­è§£é‡Šå†å²äº‹ä»¶ã€‚",
                custom_greeting="ä½ å¥½ï¼æˆ‘æ˜¯å†å²å­¦å®¶ï¼Œå¾ˆé«˜å…´ä¸ä½ æ¢è®¨å†å²è¯é¢˜ã€‚"
            ))
            if response.success:
                print(f"   è‡ªå®šä¹‰è§’è‰²è®¾ç½®æˆåŠŸ!\n")
            else:
                print(f"   è‡ªå®šä¹‰è§’è‰²è®¾ç½®å¤±è´¥: {response.message}\n")
        except Exception as e:
            print(f"   è®¾ç½®è‡ªå®šä¹‰è§’è‰²å¤±è´¥: {e}\n")
        
        # 7. ä¸è‡ªå®šä¹‰è§’è‰²å¯¹è¯
        print("7. ä¸è‡ªå®šä¹‰è§’è‰²å¯¹è¯")
        try:
            response = stub.ChatWithRole(role_play_pb2.ChatRequest(
                message="èƒ½è®²è®²å”æœçš„å¼€å…ƒç››ä¸–å—ï¼Ÿ"
            ))
            if response.success:
                print(f"   AIå›å¤: {response.response[:100]}...\n")  # åªæ˜¾ç¤ºå‰100ä¸ªå­—ç¬¦
            else:
                print(f"   å¯¹è¯å¤±è´¥: {response.error}\n")
        except Exception as e:
            print(f"   å¯¹è¯å¤±è´¥: {e}\n")
        
        print("=== æ¼”ç¤ºå®Œæˆ ===")

if __name__ == '__main__':
    demo_all_features()